protostar/starknet/cheatable_cached_state.py

```diff
@@ -156,23 +165,46 @@ def get_abi_from_contract_address(self, contract_address: int) -> AbiType:
         return self.class_hash_to_contract_abi_map[class_hash]
 
     def register_expected_call(
-        self, contract_address: Address, selector: SelectorType, calldata: list[int]
+        self,
+        contract_address: Address,
+        function_selector: SelectorType,
+        calldata: CairoData,
     ):
         if self.expected_contract_calls.get(contract_address):
-            self.expected_contract_calls[contract_address].append((selector, calldata))
+            self.expected_contract_calls[contract_address].append(
+                (function_selector, calldata)
+            )
         else:
-            self.expected_contract_calls[contract_address] = [(selector, calldata)]
+            self.expected_contract_calls[contract_address] = [
+                (function_selector, calldata)
+            ]
 
     def unregister_expected_call(
-        self, contract_address: Address, calldata: tuple[int, list[int]]
+        self,
+        contract_address: Address,
+        function_selector: SelectorType,
+        calldata: CairoData,
     ):
-        data_for_address = self.expected_contract_calls.get(contract_address)
-        if data_for_address is not None and calldata in data_for_address:
-            for index, (selector, calldata_item) in enumerate(data_for_address):
-                if selector == calldata[0] and calldata_item == calldata[1]:
-                    del data_for_address[index]
-            if not data_for_address:
-                del self.expected_contract_calls[contract_address]
+        if contract_address not in self.expected_contract_calls:
+            return
+        self.expected_contract_calls[contract_address] = [
+            (stored_function_name, stored_calldata)
+            for (
+                stored_function_name,
+                stored_calldata,
+            ) in self.expected_contract_calls[contract_address]
+            if stored_function_name != function_selector or stored_calldata != calldata
+        ]
+        if not self.expected_contract_calls[contract_address]:
+            del self.expected_contract_calls[contract_address]
+
+    async def get_contract_class_by_address(
+        self, contract_address: Union[Address, int]
+    ) -> ContractClass:
+        if isinstance(contract_address, Address):
+            contract_address = int(contract_address)
+        class_hash = await self.get_class_hash_at(contract_address)
+        return await self.get_contract_class(class_hash=class_hash)
```

Comment:
Let's move this to the cheater itself. The cheater will have access to the state anyway

Reply:
This is also used by the cheatable syscall handler and the old arch implementation of the cheat code. If we move this to the cheater it will become exclusive to it and the old cheat code will no longer work :/

Reply:
New arch will have a separate state class, so we can leverage that by leaving the old one as-is, and leaving new implementation cleaner.

Reply:
With the new changes, it is moved to the cheater now.